NXP add scmi nxp irq mask interface

yongxu-wang15 opened the pull request:
1. Add scmi_cpu_set_irq_mask api for scmi nxp imx vendor, its aim to set cmc irq mask for nvic
2. verified it passed in imx95 m7

yongxu-wang15 committed:
drivers: firmware: scmi: add cpu irq wake interface for nxp vendor

Add scmi_cpu_set_irq_mask function for scmi nxp cpu driver,
set cmc wakeup mask from scmi interface by this.

The CPU_IRQ_WAKE_SET is sent to the scmi platform
to determine which interrupt can wakeup the cpu platform
when the corresponding interrupt occurs

"1" means the IRQ cannot wakeup CPU platform
Each bit corresponds one-to-one with the interrupt
controller of the platform such as nvic in mcore

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
soc: nxp: imx95: set and restore wakeup irq mask in pm context

The CMC interface controls the entry and exit of the
CPU's low-power mode and the identification of the wake-up source.
To ensure the normal operation of the system's low-power timing sequence,
when transfer IDLE_RUN to IDLE_SLEEP, it is necessary to ensure
that the system is not awakened by the wake-up source during this stage.
Therefore, an IRQ MASK needs to be set on the CMC

Before the CPU enters the low power mode, a wake up mask needs to be set
according to the situation where the interrupt controller is enabled
at that time.
After the cpu exits the low power mode, resume needs to be performed

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

LaurentiuM1234 commented on soc/nxp/imx/imx9/imx95/m7/soc.c:
```diff
@@ -114,8 +115,27 @@ void pm_state_before(void)
+		cpu_irq_mask_cfg.mask[val] = wake_mask[val];
+	}
+
+	scmi_cpu_set_irq_mask(&cpu_irq_mask_cfg);
```
not sure I fully understand what this is trying to achieve? is this supposed to block out the previously enabled interrupts while the core transitions to a LP state or is there more to it? is this trying to fix an existing issue?

LaurentiuM1234 reviewed (commented):
Some comments/questions below

yongxu-wang15 commented on include/zephyr/drivers/firmware/scmi/nxp/cpu.h:
```diff
@@ -52,6 +54,23 @@ struct scmi_cpu_pd_lpm_config {
+ * @param number value to set the number of mask
+ * @param mask pointer to actually mask array value
+ *
+ * @brief Describes scmi cpu irq wake mask setting
```
sure, updated

yongxu-wang15 commented on include/zephyr/drivers/firmware/scmi/nxp/cpu.h:
```diff
@@ -93,4 +112,16 @@ int scmi_cpu_sleep_mode_set(struct scmi_cpu_sleep_mode_config *cfg);
 int scmi_cpu_pd_lpm_set(struct scmi_cpu_pd_lpm_config *cfg);
+
+/**
+ * @brief Send the SCMI_CPU_DOMAIN_MSG_CPU_IRQ_WAKE_SET command and get its reply
```
yes, done

yongxu-wang15 commented on soc/nxp/imx/imx9/imx95/m7/soc.c:
```diff
@@ -114,8 +115,27 @@ void pm_state_before(void)
+
+	/* IRQs enabled at NVIC level become GPC wake sources */
+	for (uint32_t idx = 0; idx < 8; idx++) {
+		wake_mask[idx] = ~(NVIC->ISER[idx]);
```
yes, 1 will mask wakeup interrupt in cmc side even though nvic side enabled this bit in mcore side

yongxu-wang15 commented on drivers/firmware/scmi/nxp/cpu.c:
```diff
@@ -90,3 +90,42 @@ int scmi_cpu_pd_lpm_set(struct scmi_cpu_pd_lpm_config *cfg)
 
```
updated commit, done

LaurentiuM1234 reviewed (commented):
one more suggestion below. Otherwise, good with me.

yongxu-wang15 committed:
drivers: firmware: scmi: return converted status directly

Simplify error handling by returning scmi_status_to_errno(status)
directly after message response.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented on drivers/firmware/scmi/nxp/cpu.c:
```diff
@@ -90,3 +90,40 @@ int scmi_cpu_pd_lpm_set(struct scmi_cpu_pd_lpm_config *cfg)
+		return ret;
+	}
+
+	return 0;
```
yes, scmi_status_to_errno is enough here, updated

yongxu-wang15 commented:
Updated from @LaurentiuM1234 comments:

add one commit to simplify error handling by returning scmi_status_to_errno(status) directly after message response for not need other reply buffer message interface, the scmi_status_to_errno include transfer about success and error, for this case which only need check response status, the return scmi_status_to_errno(status); is enough 

thanks

LaurentiuM1234 reviewed (approved):
TYVM for taking the time to do the `scmi_status_to_errno()` fix in all locations!

JiafeiPan commented on include/zephyr/drivers/firmware/scmi/nxp/cpu.h:
```diff
@@ -23,6 +23,8 @@
 
 #define SCMI_CPU_MAX_PDCONFIGS_T 7U
 
+#define SCMI_CPU_IRQ_WAKE_NUM	22U
```
Is this a fixed number or could be different for each platform?

yongxu-wang15 replied:
The maximum scmi payload is 100, the SCMI_CPU_IRQ_WAKE_NUM is same for all platform, which already supports the max number of bytes supported by this structure
